### PR TITLE
bazel/clang-tidy: Cleanups and optimizations

### PR DESCRIPTION
**Pull request by phlax**

### CODE REVIEW

**Comment by phlax** [comment]
cc @mathetake

File: bazel/format/clang_tidy/clang_tidy.bzl [hunk 1]
```diff
@@ -79,23 +60,25 @@ def _run_tidy(
...
+def _check_valid_file_type(src):
+    """
+    Returns True if the file type matches one of the permitted srcs file types for C and C++ header/source files.
+    """
+    permitted_file_types = [
+        ".c", ".cc", ".cpp", ".cxx", ".c++", ".C", ".h", ".hh", ".hpp", ".hxx", ".inc", ".inl", ".H",
+    ]
+    for file_type in permitted_file_types:
+        if src.basename.endswith(file_type):
+            return True
```
Comment (mathetake) [inline bazel/format/clang_tidy/clang_tidy.bzl:73]: can we use dict here? for loop with more than 10 items seems a bit unnecessary linear search

File: bazel/format/clang_tidy/run_clang_tidy.sh [hunk 2]
```diff
@@ -21,4 +21,6 @@ truncate -s 0 "$OUTPUT"
 # place it in the current directory
 test -e .clang-tidy || ln -s -f "$CONFIG" .clang-tidy
 
+# echo "RUN: ${CLANG_TIDY_BIN} ${*}"
+
```
Comment (mathetake) [inline bazel/format/clang_tidy/run_clang_tidy.sh:25]: leftover?

**Review by mathetake — APPROVED** [review]

File: bazel/format/clang_tidy/clang_tidy.bzl (hunk 1 above)
Comment (phlax) [reply bazel/format/clang_tidy/clang_tidy.bzl:73]: not sure i follow exactly - in real python there are various ways to optimize this but i dont think they work in starlark - can you give example?

File: bazel/format/clang_tidy/run_clang_tidy.sh (hunk 2 above)
Comment (phlax) [reply bazel/format/clang_tidy/run_clang_tidy.sh:25]: i figured keep it - its kinda useful for debugging

File: bazel/format/clang_tidy/clang_tidy.bzl (hunk 1 above)
Comment (mathetake) [reply bazel/format/clang_tidy/clang_tidy.bzl:73]:
```

PERMITTED_TYPES = {"c": True, "cpp": True}

def perimtted(file_name):
    if file_name.split(".")[-1] in PERMITTED_TYPES:
        return True
```

like this?

Comment (phlax) [reply bazel/format/clang_tidy/clang_tidy.bzl:73]: not sure if the dict would be faster than a list

what i think we can do is ...

```suggestion
    permitted_file_types = [
        ".c", ".cc", ".cpp", ".cxx", ".c++", ".C", ".h", ".hh", ".hpp", ".hxx", ".inc", ".inl", ".H",
    ]
    return src.extension in permitted_file_types
```

Comment (phlax) [reply bazel/format/clang_tidy/clang_tidy.bzl:73]: probably - we can shift the list out and get rid of the fun also

Comment (mathetake) [reply bazel/format/clang_tidy/clang_tidy.bzl:73]: at least can you move `.h` and `.cc` to the first elements?

Comment (mathetake) [reply bazel/format/clang_tidy/clang_tidy.bzl:73]: they are the most common in envoy code base

Comment (phlax) [reply bazel/format/clang_tidy/clang_tidy.bzl:73]: re `.h` i think we dont want it to run over the headers directly it causes a stack of other issues

if we _do_ want that then we need to fish them out from the `hdrs` attr

File: bazel/format/clang_tidy/clang_tidy.bzl [hunk 3]
```diff
@@ -158,11 +137,11 @@ def _clang_tidy_aspect_impl(target, ctx):
             src,
             target.label.name,
         )
-        for src in srcs
+        for src in _rule_sources(ctx)
```
Comment (mathetake) [inline bazel/format/clang_tidy/clang_tidy.bzl:140]: I think this outputs can be one run i guess? i don't see any reason why we cannot run clang-tidy for all source codes in the target at once vs here run each file per clang-tidy.

Comment (phlax) [reply bazel/format/clang_tidy/clang_tidy.bzl:130]: let me test ...

Comment (phlax) [reply bazel/format/clang_tidy/clang_tidy.bzl:130]: so the one issue with this plan is that the flags are set according to whether `src` is `.c` or `.cc`

a quick grep of the code suggests there are not any builds that have both as a src (not sure that would make sense - but what do i know)

on the upsides it makes the code cleaner and removes the need to use `to_list` on the srcs

Comment (mathetake) [reply bazel/format/clang_tidy/clang_tidy.bzl:130]: i believe all Envoy codes are C++ hence .cc and `.c` are for test codes, so we can simply ignore them

Comment (phlax) [reply bazel/format/clang_tidy/clang_tidy.bzl:130]: so next problem is 

```console
 38     outfile = ctx.actions.declare_file(
 39         "bazel_clang_tidy_%s.%s.clang-tidy.yaml" % (infile.path, discriminator)
 40     )

```

i can workaround for now -  i think in the original implementation this was intended for the fix file altho not sure it was actually implemented

my intention is to use this output for the output from clang-tidy and then collect those up so probs we can think of a different path scheme

Comment (phlax) [reply bazel/format/clang_tidy/clang_tidy.bzl:130]: lets deal with this separately - the benefit is marginal at best and it requires quite a lot of change to make it work

**Pull request merged**